scipy/spatial/hausdorff_util.h

```diff
@@ -0,0 +1,59 @@
+// Utility C code for directed_hausdorff
+// Try to improve on the performance of the original Cython directed_hausdorff
+
+#include <math.h>
+
+struct return_values 
+{
+    double cmax;
+    int    index_1;
+    int    index_2;
+};
+
+// function for inner loop of directed_hausdorff algorithm
+void hausdorff_loop(const int data_dims,
+                    const double ar1[],
+                    const double ar2[],
+                    const int N1,
+                    const int N2,
+                    struct return_values * ret_vals)
+{
+    double               d, cmin, diff;
+    const double * const ar1_start_Ptr = ar1;
+    const double * const ar2_start_Ptr = ar2;
+    const int size_ar1 = data_dims * N1;
+    const int size_ar2 = data_dims * N2;
+    const double * const ar1_end_Ptr = &ar1[size_ar1 - 1];
+    const double * const ar2_end_Ptr = &ar2[size_ar2 - 1];
+    int k;
+
+    ret_vals->cmax = 0;
+    
+    for ( ; ar1 <= ar1_end_Ptr; ar1 += data_dims ) {
+        cmin = INFINITY;
+        for ( ar2 = ar2_start_Ptr; ar2 <= ar2_end_Ptr; ar2 += data_dims) {
+            d = 0;
+            for ( k = 0; k < data_dims; ++k ) {
+                diff = *(ar1 + k) - *(ar2 + k);
```

Comment:
If data_dims is high, then you could look at vectorizing with a BLAS function. Function call overhead will be high though.